### DISCUSSION TITLE
Methods to avoid `slowDownRadius` when `requestMoveTarget`.

### DISCUSSION THREAD

**Discussion by rob-myers** [isaac-mason/recast-navigation-js#380] [Ideas]
When an agent approaches a `requestMoveTarget` position, it slows down a bit ([arrival](https://www.red3d.com/cwr/steer/gdc99/#:~:text=Arrival%20behavior%20is%20identical%20to,as%20shown%20in%20Figure%206.)) due to the following line of code:
> https://github.com/recastnavigation/recastnavigation/blob/455a019e7aef99354ac3020f04c1fe3541aa4d19/DetourCrowd/Source/DetourCrowd.cpp#L1204

```cpp
const float slowDownRadius = ag->params.radius*2;	
```

It doesn't slow down particularly quickly.

Ideally we should be able to force this, relying upon e.g. [`AnimationAction.fadeIn`
](https://threejs.org/docs/#api/en/animation/AnimationAction.fadeIn) for a smooth animated transition. I've tried two methods:
- When close to the target switch to `requestMoveVelocity`, then stop when really close.
- When close to the target `requestMoveTarget` again, overshooting a bit, then stop a bit early so you end up at the original target.

---

The first seems preferable, but I'd prefer to specify a parameter somewhere.

Could we alter the underlying C++ code e.g. somehow provide a new agent parameter?

**Comment by isaac-mason** [comment]
Hey @rob-myers!

If it's just a matter of changing wrapper javascript / c++ code here, we can for sure. If you're thinking of swapping between requestMoveVelocity / requestMoveTarget, maybe this could be done using the existing recast-navigation-js APIs? Or if you're imagining new agent slowdown parameters, maybe we could modify recast-navigation-js's Crowd javascript class and add parameters there?

If you think these changes are best made inside recastnavigation though, I think it'd be ideal to try contributing any changes to the upstream repo. You could fork both recast-navigation-js & recastnavigation to test/use your changes in the meantime. You can modify this build script to clone recastnavigation from a different repo / comment it out and link a local folder: https://github.com/isaac-mason/recast-navigation-js/blob/main/packages/recast-navigation-wasm/build.sh#L18

**Reply by rob-myers** [reply]
Hey,

I found a workaround:

> when close to the target, we can adjust the `maxSpeed` to undo the scaling done in recast-detour.

```js
// onTick method
if (distanceToTarget < 2 * agentRadius) {
  this.agent.updateParameters({ maxSpeed: this.getMaxSpeed() * ((2 * agentRadius) / distanceToTarget) });
}
```

This is preferable because `requestMoveTarget` will steer the agent, whereas `requestMoveVelocity` does not. I'd still prefer a parameter, so I guess I'll request one in the underlying repo. Cheers!

**Reply by isaac-mason** [reply]
Nice, thanks for sharing!
